### ISSUE TITLE
opam update fails with unhelpful git error

### ISSUE THREAD

**Issue by ygrek** [ocaml/opam#3485]
```
# opam config report
# opam-version      2.0.0 (b2b8f1e32c355a1230d9be9c54a2f56750d3dc7f) 
# self-upgrade      no
# system            arch=x86_64 os=linux os-distribution=debian os-version=9.4
# solver            builtin-mccs+glpk
# install-criteria  -removed,-count[version-lag,request],-count[version-lag,changed],-changed
# upgrade-criteria  -removed,-count[version-lag,solution],-new
# jobs              4
# repositories      1 (http), 1 (local), 4 (version-controlled) (default repo at 70ed52ef)
# pinned            0
# current-switch    4.03.0
```

```
$ opam update
[ERROR] Command "/usr/bin/git diff --no-ext-diff --quiet" failed:
"/usr/bin/git diff --no-ext-diff --quiet" exited with code 128 "fatal: unable to read 926b1878be3564c43af358b3362cfe83ba7abfb4"
```
This is second time I am getting this in past month and I don't know why.
This is apparently cause by git doing
`1463  12:09:48.665131 open(".git/objects/92/6b1878be3564c43af358b3362cfe83ba7abfb4", O_RDONLY|O_NOATIME|O_CLOEXEC) = -1 ENOENT (No such file or directory) <0.000007>`
but I don't understand what is this repository? it doesn't have any remotes, not any commits?
```
$ git  -C ~/.opam/download-cache/git ls
fatal: your current branch 'master' does not have any commits yet
$ git  -C ~/.opam/download-cache/git st
fatal: This operation must be run in a work tree
```
More helpful error message would be nice, answering the question "what was opam trying to do when this happened"..

Doing `opam update default` works fine.
Removing ~/.opam/download-cache doesn't help.
Creating empty git repository in ~/.opam/download-cache/git doesn't help either.
And I cannot find where that commit that it tries to diff is coming from

**Comment by ygrek** [comment]
removing ~/.opam/repo/state.cache doesn't help (state.cache gets recreated and then same error)

**Comment by ygrek** [comment]
looking at the log/*.info file I found the package with url to git that apparently was causing this. Removing that package fixed the issue. I still don't know the root cause. Better error message from opam would have really helped

**Comment by AltGr** [comment]
The message should indeed _at least_ specify during the update of _what_ it failed. It looks like you have either a repository or a local git-pinned package with broken git data?

What pointed you to `.opam/download-cache/git` ? It makes sense that most git commands don't work there, since it's a bare repository that is only used as a git object cache.

Thanks for reporting!

**Comment by ygrek** [comment]
> It looks like you have either a repository or a local git-pinned package with broken git data?

yes, I had several git pins and packages with git urls (see comment above)

> What pointed you to .opam/download-cache/git ? It makes sense that most git commands don't work there, since it's a bare repository that is only used as a git object cache.

strace
